Added search tab for communitiyes (fixes #88)

JakeShirley opened the pull request:
<!-- 
Thank you for making a pull request! 
Since we are very busy with getting Mlem into a releaseable state, we had to introduce this short questionnaire to help us review PRs.
Before you submit your PR, please take a few minutes to fill out all the needed information.

Please note that if you do not fill out the checklist, your PR will be automatically rejected unless you are an approved contributor. 
We apologize, as we would love to dedicate the time it deserves to every PR, but at present, we are under considerable time pressure.
-->

# Checklist
- [x] I have read [CONTRIBUTING.md](./CONTRIBUTING.md)
- [x] I have described what this PR contains
- [x] This PR addresses one or more open issues that were assigned to me:
      - Fixed #88 
- [x] If this PR alters the UI, I have attached pictures/videos

# Pull Request Information

## About this Pull Request
Adds a new search tab to the bottom bar.

I implemented paging on my own because I didn't think the feed tracker stuff would scale to multiple result types at the moment.  It's.... not pretty.

A few other side bits:
- Added utility to `Int` to print things like `1.2k` and `1.1m`
- Updated search tracker to use views so we can display how many subscribers communities have
- This does NOT touch the existing search, but it should be downgrade to swapping between "Subscribed", "Local", and "All" (no search)

## Screenshots and Videos
![image](https://github.com/mlemgroup/mlem/assets/1000311/9f20dec5-0071-4311-9d90-e5abbc519f8c)

![image](https://github.com/mlemgroup/mlem/assets/1000311/697e3515-a256-4840-b165-7c6a19af3b7f)


## Additional Context
**Any additional context you'd like to add to help us review this PR**

EricBAndrews commented on Mlem/Models/Trackers/RecentSearchesTracker.swift:
```diff
@@ -0,0 +1,73 @@
+                    whatToDecode: .recentSearches
+                ) as? [String] ?? []
+            } catch let decodingError {
+                print("Failed while decoding recent searches, erasing file: \(decodingError)")
```
Does the file actually get erased?

EricBAndrews commented on Mlem/Views/Tabs/Search/Search View.swift:
```diff
@@ -0,0 +1,183 @@
+        }
+        .searchable(text: getSearchTextBinding(), prompt: "Search for communities")
+        .onSubmit(of: .search) {
+            preformSearch()
```
nit:

preform -> perform

EricBAndrews commented on Mlem/Views/Tabs/Search/Search View.swift:
```diff
@@ -0,0 +1,183 @@
+    }
+    
+    @ViewBuilder
+    private var searchedContents: some View {
+        if isSearching && communitySearchResultsTracker.foundCommunities.isEmpty {
+            LoadingView(whatIsLoading: .search)
+        } else if communitySearchResultsTracker.foundCommunities.isEmpty {
+            Text("No communities found for search")
+        } else {
+            List {
+                ForEach(communitySearchResultsTracker.foundCommunities) { community in
+                    CommunityLinkView(
+                        community: community.community,
+                        extraText: "\(community.counts.subscribers.roundedWithAbbreviations) subscribers"
+                    )
+                        .frame(maxWidth: .infinity, alignment: .leading)
+                        .onAppear {
+                            let communityIndex = communitySearchResultsTracker.foundCommunities.firstIndex(of: community)
+                            if let index = communityIndex {
+                                
+                                // If we are half a page from the end, ask for more
+                                let distanceFromEnd = communitySearchResultsTracker.foundCommunities.count - index
+                                if distanceFromEnd == searchPageSize / 2 {
+                                    if hasMorePages {
+                                        preformSearch()
+                                    }
+                                }
+                            }
+                        }
+                }
+            }
+        }
+    }
```
Good for now, long-term we can look at maybe using one of those shiny generic feed trackers

EricBAndrews commented on Mlem/Views/Tabs/Search/Search View.swift:
```diff
@@ -0,0 +1,183 @@
+        }
+    }
+    
+    private func loadRecentSearchesFromDisk() {
+        
+    }
```
?

JakeShirley replied:
What? You don't like functions that do nothing and are never called??

EricBAndrews reviewed (changes requested):
Couple minor things, but looks good overall!

EricBAndrews commented on Mlem/ContentView.swift:
```diff
@@ -37,13 +37,19 @@ struct ContentView: View {
+                NavigationView {
+                    SearchView(account: currentActiveAccount)
+                } .tabItem {
+                    Label("Search", systemImage: "text.magnifyingglass")
```
Can we make this `magnifyingglass` when it's not selected and `text.magnifyingglass` when it is?

EricBAndrews commented:
> Design notes/questions:
> 
> Can we center the “clear recent searches” within that button?
> 
> Should we take it out of the button itself and just have it be text on the black background? Might be a bit cleaner that way? Thoughts?
> 
> Should the number of subscribers text remain white? Or would a gray look a little better?

I think the communities should be white and the subscribers gray--right now the weight feels backwards

JakeShirley commented:
Addressed all of the feedback other than the label coloring of names / subs.  I think we can do this but that will be a little deeper surgery for the community link view and it'd be nice to get a beta out this morning.
